File: leetcode/Remove Nth Node From End of List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @return a ListNode
    def removeNthFromEnd(self, head, n):
        myhead = ListNode(0)
        myhead.next = head
        now_head = myhead
        last_head = myhead
        p = head
        i = 0
        while p:
            i += 1
            if i == n:
                last_head = now_head
                now_head = p
                i = 0
            p = p.next
            
        if i == n:
            last_head.next = last_head.next.next
        elif last_head == myhead and i == 0:
            return head.next
        else:
            for j in range(i):
                last_head = last_head.next
            last_head.next = last_head.next.next
        return head
                    

File: lintcode/Remove_Nth_Node_From_End_of_List.cpp
/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @param n: An integer.
     * @return: The head of linked list.
     */
    ListNode *removeNthFromEnd(ListNode *head, int n) {
        if (NULL == head || 0 == n) return NULL;
        
        ListNode dummy(0);
        dummy.next = head;
        ListNode *slow = &dummy;
        ListNode *fast = &dummy;
        int k = n;
        while (k > 0 && fast->next) {
            fast = fast->next;
            k--;
        }
        if (k != 0) return NULL;
        
        while (fast && fast->next) {
            fast = fast->next;
            slow = slow->next;
        }
        
        ListNode *tmp = slow->next;
        slow->next = slow->next->next;
        delete tmp;
        return dummy.next;
    }
};



